@smit-hinsu @gbaned Looks like the build failure is unrelated and someone else broke it.

```
 /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/k8-opt/bin/external/llvm-project/llvm/_objs/PowerPCCodeGen/PPCAsmPrinter.pic.d '-frandom-seed=bazel-out/k8-opt/bin/external/llvm-project/llvm/_objs/PowerPCCodeGen/PPCAsmPrinter.pic.o' -fPIC -DLLVM_ENABLE_STATS -D__STDC_LIMIT_MACROS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DLLVM_BUILD_GLOBAL_ISEL -iquote external/llvm-project -iquote bazel-out/k8-opt/bin/external/llvm-project -iquote external/zlib -iquote bazel-out/k8-opt/bin/external/zlib -isystem external/llvm-project/llvm/include -isystem bazel-out/k8-opt/bin/external/llvm-project/llvm/include -isystem external/zlib -isystem bazel-out/k8-opt/bin/external/zlib -isystem external/llvm-project/llvm/include/llvm/IR -isystem bazel-out/k8-opt/bin/external/llvm-project/llvm/include/llvm/IR -isystem external/llvm-project/llvm/lib/Target/AMDGPU -isystem bazel-out/k8-opt/bin/external/llvm-project/llvm/lib/Target/AMDGPU -isystem external/llvm-project/llvm/lib/Target/PowerPC -isystem bazel-out/k8-opt/bin/external/llvm-project/llvm/lib/Target/PowerPC -w -DAUTOLOAD_DYNAMIC_KERNELS -Wno-sign-compare '-std=c++14' -Iexternal/llvm-project/llvm/lib/Target/PowerPC -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/llvm-project/llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp -o bazel-out/k8-opt/bin/external/llvm-project/llvm/_objs/PowerPCCodeGen/PPCAsmPrinter.pic.o)
Execution platform: @local_execution_config_platform//:platform
external/llvm-project/llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:86:14: error: specialization of 'template<class T> struct llvm::DenseMapInfo' in different namespace [-fpermissive]
 struct llvm::DenseMapInfo<std::pair<const MCSymbol *, MCSymbolRefExpr::VariantKind>> {
              ^
In file included from external/llvm-project/llvm/include/llvm/ADT/DenseMapInfo.h:16:0,
                 from external/llvm-project/llvm/include/llvm/ADT/DenseMap.h:16,
                 from external/llvm-project/llvm/include/llvm/MC/MCAsmLayout.h:12,
                 from external/llvm-project/llvm/lib/Target/PowerPC/MCTargetDesc/PPCMCExpr.h:12,
                 from external/llvm-project/llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:19:
external/llvm-project/llvm/include/llvm/ADT/APInt.h:34:30: error:   from definition of 'template<class T> struct llvm::DenseMapInfo' [-fpermissive]
 template <typename T> struct DenseMapInfo;
                              ^
[32m[13,837 / 14,611][0m Compiling tensorflow/core/kernels/matmul_op_real.cc; 288s local ... (32 actions, 31 running)
[32m[16,863 / 18,594][0m 19 / 1198 tests, [31m[1m19 failed[0m;[0m Compiling tensorflow/compiler/mlir/tensorflow/translate/import_model.cc; 54s local ... (31 actions, 30 running)
[32m[18,505 / 18,594][0m 19 / 1198 tests, [31m[1m19 failed[0m;[0m Compiling tensorflow/core/kernels/mkl/mkl_cwise_ops_common.cc; 205s local ... (32 actions, 31 running)
```

--- a/tensorflow/compiler/mlir/tensorflow/tests/lower_tf.mlir
+++ b/tensorflow/compiler/mlir/tensorflow/tests/lower_tf.mlir
@@ -276,6 +276,16 @@ func @fourdim_space_to_batch_nd(%input: tensor<3x5x7x10xf32>, %block_shape: tens
   return %0 : tensor<?x?x?x10xf32>
 }
 
+// Verify SpaceToBatchND with input tensor of element type f16. This test case is derived from 'fourdim_space_to_batch_nd'. It checks the output
+// tensor shape and element type in a few lines in the resulting lowering.
+// CHECK-LABEL: space_to_batch_nd_element_type_f16
+func @space_to_batch_nd_element_type_f16(%input: tensor<3x5x7x10xf16>, %block_shape: tensor<2xi64>, %paddings: tensor<2x2xi64>) -> tensor<?x?x?x10xf16> {
+  // CHECK-DAG: "tf.PadV2"(%arg0, [[FULL_PADDINGS]], [[PAD_DEFAULT]]) {{.*}} -> tensor<3x?x?x10xf16>
+  // CHECK-DAG: return {{.*}}: tensor<?x?x?x10xf16>
+  %0 = "tf.SpaceToBatchND"(%input, %block_shape, %paddings) : (tensor<3x5x7x10xf16>, tensor<2xi64>, tensor<2x2xi64>) -> tensor<?x?x?x10xf16>
+  return %0 : tensor<?x?x?x10xf16>
+}
+
 // Verify the result shape for the tf.PadV2 op.
 // CHECK-LABEL: const_paddings_space_to_batch_nd
 func @const_paddings_space_to_batch_nd(%arg0: tensor<1x8x2xf32>) -> (tensor<3x5x2xf32>) {

--- a/tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc
+++ b/tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc
@@ -822,6 +822,7 @@ class LowerSpaceToBatchNDOp : public RewritePattern {
 
     Location loc = op.getLoc();
     auto input_type = op.input().getType().cast<TensorType>();
+    auto element_type = input_type.getElementType();
     if (!input_type.hasStaticShape()) {
       return failure();
     }
@@ -892,8 +893,7 @@ class LowerSpaceToBatchNDOp : public RewritePattern {
       block_shape_ints.resize(block_shape_type.getNumElements(), -1);
     }
 
-    auto padded_type =
-        RankedTensorType::get(padded_shape, rewriter.getF32Type());
+    auto padded_type = RankedTensorType::get(padded_shape, element_type);
     // padded = pad(input, full_paddings)
     auto padded =
         rewriter.create<PadOp>(loc, padded_type, op.input(), full_paddings);
@@ -971,9 +971,8 @@ class LowerSpaceToBatchNDOp : public RewritePattern {
         rewriter, loc, rewriter.getIntegerType(64), reshaped_shape_vals);
 
     auto reshaped = rewriter.create<ReshapeOp>(
-        loc,
-        RankedTensorType::get(reshaped_shape_ints, input_type.getElementType()),
-        padded, reshaped_shape);
+        loc, RankedTensorType::get(reshaped_shape_ints, element_type), padded,
+        reshaped_shape);
 
     SmallVector<int64_t, 6> permutation_vals;
     for (int64_t i = 0; i < block_rank; ++i) {

